File: Contest 50 Weekly #30/998. 构造队列/li.cpp
#include <bits/stdc++.h>
using namespace std;


// 2366 ms
class Solution {
public:
    /**
     * @param n: The array sum
     * @param arr1: The size
     * @param arr2: How many numbers small than itself
     * @return: The correct array
     */
    vector<int> getQueue(int n, vector<int> arr1, vector<int> arr2) {
        vector<vector<int>> a;
        for (int i = 0; i < n; ++i)
            a.push_back({arr1[i], arr2[i]});
        sort(a.begin(), a.end());
        vector<int> res;
        for (int i = 0; i < n; ++i)
            res.insert(res.begin() + a[i][1], a[i][0]);
        return res;
    }
};

/********************* test case *********************/
int main() {
    Solution s;
    for (int i : s.getQueue(5, {1, 2, 3, 4, 5}, {0, 0, 0, 1, 3})) cout << i << " ";
    cout << endl;
}

File: Contest 50 Weekly #30/998. 构造队列/li.py
class Solution:
    """
    @param n: The array sum
    @param arr1: The size
    @param arr2: How many numbers small than itself
    @return: The correct array
    """
    def getQueue(self, n, arr1, arr2):
        res = [0] * n
        ids = range(n)
        for v, i in sorted(zip(arr1, arr2))[::-1]:
            res[ids.pop(i)] = v
        return res


############ test cases ###########
s = Solution()
print s.getQueue(4, arr1=[1, 3, 7, 6], arr2=[0, 1, 3, 2])
